src/lib/reasoners/intervals_intf.ml

```diff
@@ -0,0 +1,735 @@
+
+  val approx_map_inc_to_set : ('a -> bnd) -> ('a -> bnd) -> 'a union -> set
+  (** [approx_map_inc_to_set f_lb f_ub u] is a variant of [map_to_set] that is
+      appropriate for (weakly) increasing functions.
+
+      The (weakly) increasing function [f] is represented by a pair
+      [(f_lb, f_ub)] of functions such that for all [x] in the domain of [f],
+      [f_lb x <= f x <= f_ub x] (and [f_ub x < f_lb x] otherwise).
+
+      Note that, unlike [partial_map_inc], this function can be used with
+      imprecise approximations: we allow [f_lb x < f x < f_ub x] (this is useful
+      for radicals). *)
+
+  val approx_map_dec_to_set : ('a -> bnd) -> ('a -> bnd) -> 'a union -> set
+  (** [approx_map_inc_to_set f_lb f_ub u] is a variant of [map_to_set] that is
+      appropriate for (weakly) decreasing functions.
+
+      The (weakly) decreasing function [f] is represented by a pair
+      [(f_lb, f_ub)] of functions such that for all [x] in the domain of [f],
+      [f_lb x <= f x <= f_ub x] (and [f_ub x < f_lb x] otherwise).
+
+      Note that, unlike [partial_map_dec], this function can be used with
+      imprecise approximations: we allow [f_lb x < f x < f_ub x]. *)
+
+  val map_inc_to_set : ('a -> bnd) -> 'a union -> set
+  (** [map_inc_to_set f u] is a variant of [approx_map_inc_to_set] when the
+      underlying function is total and precisely known.
+
+      See also {!map_strict_inc}.
+
+      Note that strict monotony is {b not} required for [map_inc_to_set]. *)
+
+  val map_dec_to_set : ('a -> bnd) -> 'a union -> set
+  (** [map_dec_to_set f u] is a variant of [approx_map_dec_to_set] when the
+      underlying function is total and precisely known.
+
+      See also {!map_strict_dec}.
+
+      Note that strict monotony is {b not} required for [map_dec_to_set]. *)
+
+  val trisection_map_to_set :
+    bnd -> t -> (t -> set) -> (unit -> set) -> (t -> set) -> set
+    (** [trisection_map_to_set v u f_lt f_eq f_gt] constructs an union of
+        intervals by combining the image of [f_lt] on the fragment of [u] that
+        only contains values strictly less than [v], the image of [f_gt] on the
+        fragment of [u] that only contains values strictly greater than [v], and
+        the image of [f_eq] if [v] is contained in [u].
+
+        It is helpful to build piecewise monotone functions such as
+        multiplication, where trisection around [0] can be used. *)
+end
+
+(** Polymorphic {!module-type-Union} implementations. *)
+module type Core = sig
+  type explanation
+  (** The type of explanations. *)
+
+  type 'a union
+  (** Normalized non-empty union of intervals over ['a] values. *)
+
+  module Union(OT : OrderedType) : Union
+    with type 'a union = 'a union
+     and type value := OT.finite
+     and type bnd = OT.t
+     and type explanation := explanation
+end
+
+module type RingType = sig
+  (** An ordered ring is an ordered type with addition and multiplication that
+      follows ring laws. *)
```

Comment:
I guess you mean `ordered ring laws` here? The ring structure has to respect the order.

Reply:
I initially meant that addition and multiplication must follow ring laws (e.g. distributivity etc.) but you are right, this should follow ordered ring laws (including compatibility with order).